File: trunk/Source/BasicAgents/ClientBase/ClientBase.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Net.Sockets;
using System.IO;
using GridSoccer.Common;
using System.Threading;

namespace GridSoccer.ClientBasic
{
    public abstract class ClientBase
    {
        private TcpClient socket = null;
        private BinaryWriter binWriter = null;
        private BinaryReader binReader = null;

        protected int m_myIndex = -1;
        private int m_goalUpperRow = -1;
        private int m_goalLowerRow = -1;

        public Sides MySide { get; private set; }
        public string MyTeamName { get; private set; }
        public int MyUnum { get; private set; }

        public int OurScore { get; private set; }
        public int OppScore { get; private set; }

        public int Cycle { get; private set; }
        public int LastSeeBall { get; private set; }
        public Position BallPosition { get; private set; }
        public bool IsGameStarted { get; private set; }
        private int CycleLength { get; set; }
        private bool JetMode { get; set; }

        public int[] LastSeePlayers { get; private set; }
        public Position[] PlayerPositions { get; private set; }
        public bool[] PlayerAvailabilities { get; private set; }

        public int EnvRows { get; private set; }
        public int EnvCols { get; private set; }
        public int EnvPassDistance { get; private set; }
        public int EnvVisibilityDistance { get; private set; }
        public int EnvGoalWidth { get; private set; }
        public int EnvMinPlayers { get; private set; }
        public int EnvMaxPlayers { get; private set; }

        public ClientBase(string serverAddr, int serverPort, string teamname, int unum)
        {
            if (String.IsNullOrEmpty(teamname))
                throw new Exception("Invalid team name", null);

            socket = new TcpClient(serverAddr, serverPort);
            socket.NoDelay = true;
            binWriter = new BinaryWriter(socket.GetStream());
            binReader = new BinaryReader(socket.GetStream());

            Send(String.Format("(init {0} {1})", teamname, unum));

            string msg = Read();
            IMessageInfo mi = ServerMessageParser.ParseInputMessage(msg);
            if (mi.MessageType != MessageTypes.InitOK)
                throw new Exception("expected init-ok but received " + mi.MessageType.ToString());

            InitOKMessage okmi =  mi as InitOKMessage;
            this.MySide = okmi.Side;
            this.MyTeamName = teamname;
            this.MyUnum = unum;

            msg = Read();
            mi = ServerMessageParser.ParseInputMessage(msg);
            if (mi.MessageType != MessageTypes.Settings)
                throw new Exception("expected settings but received " + mi.MessageType.ToString());
            ParseSettings((mi as SettingsMessage).SettingsMsgTokens);
            
            BallPosition = new Position();
            LastSeePlayers = new int[2 * EnvMaxPlayers];

            PlayerPositions = new Position[2 * EnvMaxPlayers];
            for (int i = 0; i < PlayerPositions.Length; ++i)
            {
                PlayerPositions[i] = new Position();
            }

            PlayerAvailabilities = new bool[2 * EnvMaxPlayers];

            m_myIndex = GetPlayerIndex(MySide, MyUnum);
            m_goalUpperRow = CalculateGoalUpperRow();
            m_goalLowerRow = CalculateGoalLowerRow();
        }

        #region Communication Stuff
        private void ParseSettings(string[] tokens)
        {
            int value;
            for (int i = 1; i < tokens.Length; i+=2)
            {
                if (Int32.TryParse(tokens[i + 1], out value))
                {
                    switch (tokens[i])
                    {
                        case "rows":
                            this.EnvRows = value;
                            break;
                        case "cols":
                            this.EnvCols = value;
                            break;
                        case "goal-width":
                            this.EnvGoalWidth = value;
                            break;
                        case "pass-dist":
                            this.EnvPassDistance = value;
                            break;
                        case "visible-dist":
                            this.EnvVisibilityDistance = value;
                            break;
                        case "min-players":
                            this.EnvMinPlayers = value;
                            break;
                        case "max-players":
                            this.EnvMaxPlayers = value;
                            break;
                    }
                }
            }
        }

        private void ParseSeeMessage(string[] toks)
        {
            int cycle = Int32.Parse(toks[1]);
            this.Cycle = cycle;
            int unum, num1, num2;
            int pi;

            for (int i = 2; i < toks.Length; i += 3)
            {
                switch (toks[i])
                {
                    case "score":
                        num1 = Int32.Parse(toks[i + 1]);
                        num2 = Int32.Parse(toks[i + 2]);
                        if (MySide == Sides.Left)
                        {
                            this.OurScore = num1;
                            this.OppScore = num2;
                        }
                        else
                        {
                            this.OurScore = num2;
                            this.OppScore = num1;
                        }
                        break;
                    case "self":
                        num1 = Int32.Parse(toks[i + 1]);
                        num2 = Int32.Parse(toks[i + 2]);
                        this.PlayerPositions[m_myIndex].Set(num1, num2);
                        this.LastSeePlayers[m_myIndex] = cycle;
                        this.PlayerAvailabilities[m_myIndex] = true;
                        break;
                    case "b":
                        num1 = Int32.Parse(toks[i + 1]);
                        num2 = Int32.Parse(toks[i + 2]);
                        this.LastSeeBall = cycle;
                        this.BallPosition.Set(num1, num2);
                        break;
                    case "l":
                        unum = Int32.Parse(toks[i + 1]);
                        num1 = Int32.Parse(toks[i + 2]);
                        num2 = Int32.Parse(toks[i + 3]);
                        pi = GetPlayerIndex(Sides.Left, unum);
                        this.PlayerPositions[pi].Set(num1, num2);
                        this.LastSeePlayers[pi] = cycle;
                        this.PlayerAvailabilities[pi] = true;
                        i++;
                        break;
                    case "r":
                        unum = Int32.Parse(toks[i + 1]);
                        num1 = Int32.Parse(toks[i + 2]);
                        num2 = Int32.Parse(toks[i + 3]);
                        pi = GetPlayerIndex(Sides.Right, unum);
                        this.PlayerPositions[pi].Set(num1, num2);
                        this.LastSeePlayers[pi] = cycle;
                        this.PlayerAvailabilities[pi] = true;
                        i++;
                        break;
                    default:
                        break;
                }
            }
        }

        private void Send(string msg)
        {
            try
            {
                binWriter.Write(msg);
            }
            catch
            {
            }
        }

        private string Read()
        {
            return binReader.ReadString();
        }

        private bool DataAvailable()
        {
            return socket.GetStream().DataAvailable;
        }

        private void SendAction(ActionTypes acType)
        {
            switch (acType)
            {
                case ActionTypes.Hold:
                    Send("(hold)");
                    break;
                case ActionTypes.MoveEast:
                    Send("(move east)");
                    break;
                case ActionTypes.MoveSouth:
                    Send("(move south)");
                    break;
                case ActionTypes.MoveWest:
                    Send("(move west)");
                    break;
                case ActionTypes.MoveNorth:
                    Send("(move north)");
                    break;
                case ActionTypes.MoveNorthEast:
                    Send("(move north-east)");
                    break;
                case ActionTypes.MoveSouthEast:
                    Send("(move south-east)");
                    break;
                case ActionTypes.MoveSouthWest:
                    Send("(move south-west)");
                    break;
                case ActionTypes.MoveNorthWest:
                    Send("(move north-west)");
                    break;
                case ActionTypes.Pass:
                    throw new Exception("Action Type cannot be Pass");
                default:
                    break;
            }
        }

        private void SendAction(SoccerAction act)
        {
            if (act.ActionType != ActionTypes.Pass)
                SendAction(act.ActionType);
            else
                Send(String.Format("(pass {0})", act.DestinationUnum));
        }

        private bool UpdateFromServer()
        {
            if (DataAvailable())
            {
                string msg = Read();
                IMessageInfo imi = ServerMessageParser.ParseInputMessage(msg);
                switch (imi.MessageType)
                {
                    case MessageTypes.Error:
                        Console.WriteLine("Error: " + msg);
                        return false;
                    case MessageTypes.See:
                        ParseSeeMessage((imi as SeeMessage).SeeMsgTokens);
                        return true;
                    case MessageTypes.Start:
                        IsGameStarted = true;
                        return false;
                    case MessageTypes.Stop:
                        gameIsNotStoped = false;
                        return false;
                    case MessageTypes.Cycle:
                        CycleLength = (imi as CycleMessage).CycleLength;
                        return false;
                    case MessageTypes.Jet:
                        JetMode = (imi as JetMessage).JetOn;
                        return false;
                }
            }
            return false;
        }

        protected void SetHomePosition(Position homePos)
        {
            SetHomePosition(homePos.Row, homePos.Col);
        }

        protected void SetHomePosition(int r, int c)
        {
            Send(String.Format("(home {0} {1})", r, c));
        }

        #endregion

        protected abstract SoccerAction Think();

        private bool gameIsNotStoped = true;
        public void Start()
        {
            while (gameIsNotStoped)
            {
                if(UpdateFromServer())
                    ThinkBase();
                //if(!JetMode)
                Thread.Sleep(1);
            }
            OnGameStopped();
        }

        virtual public void OnGameStopped()
        {
            // Nothing
        }

        private void ThinkBase()
        {
            if(IsGameStarted)
                SendAction(Think());
        }

        #region Utilities

        protected IEnumerable<int> GetAvailableTeammatesUnums()
        {
            int start = this.MySide == Sides.Left ? 0 : this.EnvMaxPlayers;
            int end = this.MySide == Sides.Left ? this.EnvMaxPlayers - 1 : 2 * this.EnvMaxPlayers - 1;
            for (int i = start; i <= end; ++i)
            {
                if (i != m_myIndex && PlayerAvailabilities[i])
                    yield return i - start + 1;
            }
        }

        protected IEnumerable<int> GetAvailableTeammatesIndeces()
        {
            int start = this.MySide == Sides.Left ? 0 : this.EnvMaxPlayers;
            int end = this.MySide == Sides.Left ? this.EnvMaxPlayers - 1 : 2*this.EnvMaxPlayers - 1;
            for (int i = start; i <= end; ++i)
            {
                if (i != m_myIndex && PlayerAvailabilities[i])
                    yield return i;
            }
        }

        protected IEnumerable<int> GetAvailableOpponentsIndeces()
        {
            int start = this.MySide != Sides.Left ? 0 : this.EnvMaxPlayers;
            int end = this.MySide != Sides.Left ? this.EnvMaxPlayers - 1 : 2 * this.EnvMaxPlayers - 1;
            for (int i = start; i <= end; ++i)
            {
                if (PlayerAvailabilities[i])
                    yield return i;
            }
        }

        protected IEnumerable<int> GetAvailableOpponentsUnums()
        {
            int start = this.MySide != Sides.Left ? 0 : this.EnvMaxPlayers;
            int end = this.MySide != Sides.Left ? this.EnvMaxPlayers - 1 : 2 * this.EnvMaxPlayers - 1;
            for (int i = start; i <= end; ++i)
            {
                if (PlayerAvailabilities[i])
                    yield return i - start + 1;
            }
        }

        protected int GetPlayerIndex(Sides side, int unum)
        {
            if (side == Sides.Left)
            {
                return unum - 1;
            }
            else
            {
                return EnvMaxPlayers + unum - 1;
            }
        }

        protected int GetPlayerUnumFromIndex(int index)
        {
            if (0 <= index && index < EnvMaxPlayers)
                return index + 1;
            else if (index < 2 * EnvMaxPlayers)
                return index - EnvMaxPlayers + 1;
            else
                return -1;
        }

        protected Position GetMyPosition()
        {
            return PlayerPositions[m_myIndex];
        }

        protected bool AmIBallOwner()
        {
            return (BallPosition == PlayerPositions[m_myIndex] && LastSeeBall == this.Cycle) ;
        }

        protected bool AreWeBallOwner()
        {
            int bi = GetBallOwnerPlayerIndex();
            int start = this.MySide == Sides.Left ? 0 : this.EnvMaxPlayers;
            int end = this.MySide == Sides.Left ? this.EnvMaxPlayers - 1 : 2*this.EnvMaxPlayers - 1;
            return (start <= bi && bi <= end);
        }

        protected int GetBallOwnerPlayerIndex()
        {
            for (int i = 0; i < PlayerAvailabilities.Length; ++i)
            {
                if (PlayerAvailabilities[i] && BallPosition == PlayerPositions[i] && LastSeePlayers[i] == LastSeeBall)
                    return i;
            }

            return -1;
        }

        private int CalculateGoalUpperRow()
        {
            return (EnvRows - EnvGoalWidth) / 2 + 1;
        }

        private int CalculateGoalLowerRow()
        {
            return CalculateGoalUpperRow() + EnvGoalWidth - 1;
        }

        protected int GoalUpperRow
        {
            get
            {
                return m_goalUpperRow;
            }
        }

        protected int GoalLowerRow
        {
            get
            {
                return m_goalLowerRow;
            }
        }

        #endregion
    }
}

File: Source/JavaSampleClient/src/jsampleclient/ClientBase.java
// Copyright (c) 2009 - 2011 
//  - Sina Iravanian <sina@sinairv.com>
//  - Sahar Araghi   <sahar_araghi@aut.ac.ir>
//
// This source file(s) may be redistributed, altered and customized
// by any means PROVIDING the authors name and all copyright
// notices remain intact.
// THIS SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED. USE IT AT YOUR OWN RISK. THE AUTHOR ACCEPTS NO
// LIABILITY FOR ANY DATA DAMAGE/LOSS THAT THIS PRODUCT MAY CAUSE.
//-----------------------------------------------------------------------

package jsampleclient;

import java.io.IOException;
import java.net.Socket;
import java.util.ArrayList;
import java.util.StringTokenizer;

public abstract class ClientBase
{
    private ArrayList<String> m_qReadMessages;
    private Socket socket = null;

    protected int m_myIndex = -1;
    private int m_goalUpperRow = -1;
    private int m_goalLowerRow = -1;
    
    protected  Sides MySide;
    protected  String MyTeamName;
    protected  int MyUnum;

    protected  int OurScore;
    protected  int OppScore;

    protected  int Cycle;
    protected  int LastSeeBall;

    protected  Position BallPosition;
    protected  boolean IsGameStarted;
    protected  int CycleLength;
    protected  boolean TurboMode;

    protected  int[] LastSeePlayers;
    protected  Position[] PlayerPositions;
    protected  boolean[] PlayerAvailabilities;

    protected  int EnvRows;
    protected  int EnvCols;
    protected  int EnvPassDistance;
    protected  int EnvVisibilityDistance;
    protected  int EnvGoalWidth; 
    protected  int EnvMinPlayers;
    protected  int EnvMaxPlayers;

    public ClientBase(String serverAddr, int serverPort, String teamname, int unum) throws Exception
    {
        if (teamname == null || teamname.length() <= 0)
            throw new Exception("Invalid team name");

        m_qReadMessages = new ArrayList<String>();
        socket = new Socket(serverAddr, serverPort);
        socket.setTcpNoDelay(true);

        Send(String.format("(init %s %d)", teamname, unum));

        String msg = Read();
        IMessageInfo mi = ServerMessageParser.ParseInputMessage(msg);
        if (mi.MessageType != MessageTypes.InitOK)
            throw new Exception("expected init-ok but received " + mi.MessageType.toString());

        InitOKMessage okmi =  (InitOKMessage)mi;
        this.MySide = okmi.Side;
        this.MyTeamName = teamname;
        this.MyUnum = unum;

        msg = Read();
        mi = ServerMessageParser.ParseInputMessage(msg);
        if (mi.MessageType != MessageTypes.Settings)
            throw new Exception("expected settings but received " + mi.MessageType.toString());
        ParseSettings(((SettingsMessage)mi).SettingsMsgTokens);
        
        BallPosition = new Position();
        LastSeePlayers = new int[2 * getEnvMaxPlayers()];

        PlayerPositions = new Position[2 * getEnvMaxPlayers()];
        for (int i = 0; i < getPlayerPositions().length; ++i)
        {
            getPlayerPositions()[i] = new Position();
        }

        PlayerAvailabilities = new boolean[2 * getEnvMaxPlayers()];

        m_myIndex = GetPlayerIndex(getMySide(), getMyUnum());
        m_goalUpperRow = CalculateGoalUpperRow();
        m_goalLowerRow = CalculateGoalLowerRow();
    }

    private void ParseSettings(String[] tokens)
    {
        int value;
        for (int i = 1; i < tokens.length; i+=2)
        {
            try
            {
                value = Integer.parseInt(tokens[i + 1]);
                if(tokens[i].equals("rows"))
                {
                    this.EnvRows = value;
                }
                else if(tokens[i].equals("cols"))
                {
                    this.EnvCols = value;
                }
                else if(tokens[i].equals("goal-width"))
                {
                    this.EnvGoalWidth = value;
                }
                else if(tokens[i].equals("pass-dist"))
                {
                    this.EnvPassDistance = value;
                }
                else if(tokens[i].equals("visible-dist"))
                {
                    this.EnvVisibilityDistance = value;
                }
                else if(tokens[i].equals("min-players"))
                {
                    this.EnvMinPlayers = value;
                }
                else if(tokens[i].equals("max-players"))
                {
                    this.EnvMaxPlayers = value;
                }
            }
            catch(Exception ex)
            {
            }
        }
    }

    private void ParseSeeMessage(String[] toks)
    {
        int cycle = Integer.parseInt(toks[1]);
        this.Cycle = cycle;
        int unum, num1, num2;
        int pi;

        for (int i = 2; i < toks.length; i += 3)
        {
            if(toks[i].equals("score"))
            {
                num1 = Integer.parseInt(toks[i + 1]);
                num2 = Integer.parseInt(toks[i + 2]);
                if (getMySide() == Sides.Left)
                {
                    this.OurScore = num1;
                    this.OppScore = num2;
                }
                else
                {
                    this.OurScore = num2;
                    this.OppScore = num1;
                }
            }
            else if(toks[i].equals("self"))
            {
                num1 = Integer.parseInt(toks[i + 1]);
                num2 = Integer.parseInt(toks[i + 2]);
                this.getPlayerPositions()[getMyIndex()].Set(num1, num2);
                this.getLastSeePlayers()[getMyIndex()] = cycle;
                this.getPlayerAvailabilities()[getMyIndex()] = true;
            }
            else if(toks[i].equals("b"))
            {
                num1 = Integer.parseInt(toks[i + 1]);
                num2 = Integer.parseInt(toks[i + 2]);
                this.LastSeeBall = cycle;
                this.getBallPosition().Set(num1, num2);
            }
            else if(toks[i].equals("l"))
            {
                unum = Integer.parseInt(toks[i + 1]);
                num1 = Integer.parseInt(toks[i + 2]);
                num2 = Integer.parseInt(toks[i + 3]);
                pi = GetPlayerIndex(Sides.Left, unum);
                this.getPlayerPositions()[pi].Set(num1, num2);
                this.getLastSeePlayers()[pi] = cycle;
                this.getPlayerAvailabilities()[pi] = true;
                i++;
            }
            else if(toks[i].equals("r"))
            {
                unum = Integer.parseInt(toks[i + 1]);
                num1 = Integer.parseInt(toks[i + 2]);
                num2 = Integer.parseInt(toks[i + 3]);
                pi = GetPlayerIndex(Sides.Right, unum);
                this.getPlayerPositions()[pi].Set(num1, num2);
                this.getLastSeePlayers()[pi] = cycle;
                this.getPlayerAvailabilities()[pi] = true;
                i++;
            }
        }
    }

    private void Send(String msg)
    {
        try
        {
            msg = msg + '\0';
            socket.getOutputStream().write(msg.getBytes());
            socket.getOutputStream().flush();
        }
        catch(Exception ex)
        {
            System.err.println("Could not send message: " + msg);
        }
    }

    private String Read() throws IOException
    {
        if(m_qReadMessages.size() > 0)
        {
            String msg = m_qReadMessages.get(0);
            m_qReadMessages.remove(0);
            return msg;
        }
        else
        {
            byte[] inputBuffer = new byte[1024];
            socket.getInputStream().read(inputBuffer);
            String msg = new String(inputBuffer);
            //System.out.println("Read: " + msg);
        
            StringTokenizer toker = new StringTokenizer(msg, "\0");
            while(toker.hasMoreTokens())
            {
                String curTok = toker.nextToken();
                //System.out.println("Tok: " + curTok);
                m_qReadMessages.add(curTok);
            }
            
            String curmsg = m_qReadMessages.get(0);
            m_qReadMessages.remove(0);
            return curmsg;
        }
    }

    public void SendAction(ActionTypes acType) throws Exception
    {
        switch (acType)
        {
            case Hold:
                Send("(hold)");
                break;
            case MoveEast:
                Send("(move east)");
                break;
            case MoveSouth:
                Send("(move south)");
                break;
            case MoveWest:
                Send("(move west)");
                break;
            case MoveNorth:
                Send("(move north)");
                break;
            case MoveNorthEast:
                Send("(move north-east)");
                break;
            case MoveSouthEast:
                Send("(move south-east)");
                break;
            case MoveSouthWest:
                Send("(move south-west)");
                break;
            case MoveNorthWest:
                Send("(move north-west)");
                break;
            case Pass:
                throw new Exception("Action Type cannot be Pass");
            default:
                break;
        }
    }

    public void SendAction(SoccerAction act) throws Exception
    {
        if (act.ActionType != ActionTypes.Pass)
            SendAction(act.ActionType);
        else
            Send(String.format("(pass %d)", act.DestinationUnum));
    }

    public boolean UpdateFromServer() throws Exception
    {
        String msg = Read();
        IMessageInfo imi = ServerMessageParser.ParseInputMessage(msg);
        switch (imi.MessageType)
        {
            case Error:
                System.out.println("Error: " + msg);
                return false;
            case See:
                ParseSeeMessage(((SeeMessage)imi).SeeMsgTokens);
                return true;
            case Start:
                IsGameStarted = true;
                return false;
            case Stop:
                isNotGameStopped = false;
                return false;
            case Cycle:
                CycleLength = ((CycleMessage)imi).CycleLength;
                return false;
            case Turbo:
                TurboMode = ((TurboMessage)imi).TurboOn;
                return false;
        }
        return false;
    }

    public void SetHomePosition(Position homePos)
    {
        SetHomePosition(homePos.Row, homePos.Col);
    }

    public void SetHomePosition(int r, int c)
    {
        Send(String.format("(home %d %d)", r, c));
    }

    protected abstract SoccerAction Think();

    private boolean isNotGameStopped = true;
    public void Start() throws Exception
    {
        while (getIsNotGameStopped())
        {
            if(UpdateFromServer())
                ThinkBase();
            //if(!TurboMode)
            //Thread.sleep(1);
        }
        OnGameStopped();
    }

    public void OnGameStopped()
    {
        // Nothing
    }

    private void ThinkBase() throws Exception
    {
        if(getIsGameStarted())
            SendAction(Think());
    }

    public void EpisodeTimeoutOurFail()
    {
        Send("(episode-timeout our-fail)");
    }

    public void EpisodeTimeoutOurPass()
    {
        Send("(episode-timeout our-pass)");
    }

    public void EpisodeTimeoutOppFail()
    {
        Send("(episode-timeout opp-fail)");
    }

    public void EpisodeTimeoutOppPass()
    {
        Send("(episode-timeout opp-pass)");
    }


    public ArrayList<Integer> GetAvailableTeammatesUnums()
    {
        ArrayList<Integer> rtValue = new ArrayList<Integer>();
        
        int start = this.getMySide() == Sides.Left ? 0 : this.getEnvMaxPlayers();
        int end = this.getMySide() == Sides.Left ? this.getEnvMaxPlayers() - 1 : 2 * this.getEnvMaxPlayers() - 1;
        for (int i = start; i <= end; ++i)
        {
            if (i != getMyIndex() && getPlayerAvailabilities()[i])
                rtValue.add(i - start + 1);
        }
        
        return rtValue;
    }

    public ArrayList<Integer> GetAvailableTeammatesIndeces()
    {
        ArrayList<Integer> rtValue = new ArrayList<Integer>();
        
        int start = this.getMySide() == Sides.Left ? 0 : this.getEnvMaxPlayers();
        int end = this.getMySide() == Sides.Left ? this.getEnvMaxPlayers() - 1 : 2* this.getEnvMaxPlayers() - 1;
        for (int i = start; i <= end; ++i)
        {
            if (i != getMyIndex() && getPlayerAvailabilities()[i])
                rtValue.add(i);
        }
        
        return rtValue;
    }

    public ArrayList<Integer> GetAvailableOpponentsIndeces()
    {
        ArrayList<Integer> rtValue = new ArrayList<Integer>();

        int start = this.getMySide() != Sides.Left ? 0 : this.getEnvMaxPlayers();
        int end = this.getMySide() != Sides.Left ? this.getEnvMaxPlayers() - 1 : 2 * this.getEnvMaxPlayers() - 1;
        for (int i = start; i <= end; ++i)
        {
            if (getPlayerAvailabilities()[i])
                rtValue.add(i);
        }
        
        return rtValue;
    }

    public ArrayList<Integer> GetAvailableOpponentsUnums()
    {
        ArrayList<Integer> rtValue = new ArrayList<Integer>();

        int start = this.getMySide() != Sides.Left ? 0 : this.getEnvMaxPlayers();
        int end = this.getMySide() != Sides.Left ? this.getEnvMaxPlayers() - 1 : 2 * this.getEnvMaxPlayers() - 1;
        for (int i = start; i <= end; ++i)
        {
            if (getPlayerAvailabilities()[i])
                rtValue.add(i - start + 1);
        }
        
        return rtValue;
    }

    public int GetPlayerIndex(Sides side, int unum)
    {
        if (side == Sides.Left)
        {
            return unum - 1;
        }
        else
        {
            return getEnvMaxPlayers() + unum - 1;
        }
    }

    public int GetPlayerUnumFromIndex(int index)
    {
        if (0 <= index && index < getEnvMaxPlayers())
            return index + 1;
        else if (index < 2 * getEnvMaxPlayers())
            return index - getEnvMaxPlayers() + 1;
        else
            return -1;
    }

    public Position GetMyPosition()
    {
        return getPlayerPositions()[getMyIndex()];
    }

    public boolean AmIBallOwner()
    {
        return (getBallPosition() == getPlayerPositions()[getMyIndex()] && getLastSeeBall() == this.getCycle()) ;
    }

    public boolean AreWeBallOwner()
    {
        int bi = GetBallOwnerPlayerIndex();
        int start = this.getMySide() == Sides.Left ? 0 : this.getEnvMaxPlayers();
        int end = this.getMySide() == Sides.Left ? this.getEnvMaxPlayers() - 1 : 2* this.getEnvMaxPlayers() - 1;
        return (start <= bi && bi <= end);
    }

    public int GetBallOwnerPlayerIndex()
    {
        for (int i = 0; i < getPlayerAvailabilities().length; ++i)
        {
            if (getPlayerAvailabilities()[i] && getBallPosition() == getPlayerPositions()[i] && getLastSeePlayers()[i] == getLastSeeBall())
                return i;
        }

        return -1;
    }

    public int CalculateGoalUpperRow()
    {
        return (getEnvRows() - getEnvGoalWidth()) / 2 + 1;
    }

    public int CalculateGoalLowerRow()
    {
        return CalculateGoalUpperRow() + getEnvGoalWidth() - 1;
    }

    public int getGoalUpperRow()
    {
        return m_goalUpperRow;
    }

    public int getGoalLowerRow()
    {
        return m_goalLowerRow;
    }

    public Sides getMySide() {
        return MySide;
    }

    public String getMyTeamName() {
        return MyTeamName;
    }

    public int getMyUnum() {
        return MyUnum;
    }

    public int getOurScore() {
        return OurScore;
    }

    public int getOppScore() {
        return OppScore;
    }

    public int getCycle() {
        return Cycle;
    }

    public int getLastSeeBall() {
        return LastSeeBall;
    }

    public Position getBallPosition() {
        return BallPosition;
    }

    public boolean getIsGameStarted() {
        return IsGameStarted;
    }

    public int getCycleLength() {
        return CycleLength;
    }

    public boolean isTurboMode() {
        return TurboMode;
    }

    public int[] getLastSeePlayers() {
        return LastSeePlayers;
    }

    public Position[] getPlayerPositions() {
        return PlayerPositions;
    }

    public boolean[] getPlayerAvailabilities() {
        return PlayerAvailabilities;
    }

    public int getEnvRows() {
        return EnvRows;
    }

    public int getEnvCols() {
        return EnvCols;
    }

    public int getEnvPassDistance() {
        return EnvPassDistance;
    }

    public int getEnvVisibilityDistance() {
        return EnvVisibilityDistance;
    }

    public int getEnvGoalWidth() {
        return EnvGoalWidth;
    }

    public int getEnvMinPlayers() {
        return EnvMinPlayers;
    }

    public int getEnvMaxPlayers() {
        return EnvMaxPlayers;
    }

    public boolean getIsNotGameStopped() {
        return isNotGameStopped;
    }

    public int getMyIndex() {
        return m_myIndex;
    }
}
